[docs] Fix EmbeddingBag docs

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45763 [docs] Fix EmbeddingBag docs**

**Summary**
This commit updates the documentation for `EmbeddingBag` to say that for
bags of constant length with no per-sample weights, the class is
equivalent to `Embedding` followed by `torch.sum(dim=1)`. The current
docs say `dim=0` and this is readily falsifiable.

**Test Plan**
1) Tried `Embedding` + `sum` with `dim`=0,1 in interpreter and compared
to `EmbeddingBag`
```
>>> import torch                                                                                     
>>> weights = torch.nn.Parameter(torch.randn(10, 3))              
>>> e = torch.nn.Embedding(10, 3)                                                                    
>>> eb = torch.nn.EmbeddingBag(10, 3, mode="sum")                                                    
>>> e.weight = weights                                                                               
>>> eb.weight = weights                           
# Use 2D inputs because we are trying to test the case in which bags have constant length                                               
>>> inputs = torch.LongTensor([[4,1,2,7],[5,6,0,3]])
>>> eb(inputs)
tensor([[-2.5497, -0.1556, -0.5166],
        [ 2.2528, -0.3627,  2.5822]], grad_fn=<EmbeddingBagBackward>)
>>> torch.sum(e(inputs), dim=0)
tensor([[ 1.6181, -0.8739,  0.8168],
        [ 0.0295,  2.3274,  1.2558],
        [-0.7958, -0.4228,  0.5961],
        [-1.1487, -1.5490, -0.6031]], grad_fn=<SumBackward1>)
>>> torch.sum(e(inputs), dim=1)
tensor([[-2.5497, -0.1556, -0.5166],
        [ 2.2528, -0.3627,  2.5822]], grad_fn=<SumBackward1>)
```
So clearly `torch.sum` with `dim=0` is not correct here.

2) Built docs and viewed in browser.

*Before*
<img width="882" alt="Captura de Pantalla 2020-10-02 a la(s) 12 26 20 p  m" src="https://user-images.githubusercontent.com/4392003/94963035-557be100-04ac-11eb-986c-088965ac3050.png">

*After*
<img width="901" alt="Captura de Pantalla 2020-10-05 a la(s) 11 26 51 a  m" src="https://user-images.githubusercontent.com/4392003/95117732-ea294d80-06fd-11eb-9d6b-9b4e6c805cd0.png">


**Fixes**
This commit closes #43197.

Differential Revision: [D24118206](https://our.internmc.facebook.com/intern/diff/D24118206)

Commit:
[docs] Fix EmbeddingBag docs

**Summary**
This commit updates the documentation for `EmbeddingBag` to say that for
bags of constant length with no per-sample weights, the class is
equivalent to `Embedding` followed by `torch.sum(dim=1)`. The current
docs say `dim=0` and this is readily falsifiable.

**Test Plan**
1) Tried `Embedding` + `sum` with `dim`=0,1 in interpreter and compared
to `EmbeddingBag`
2) Built docs and viewed in browser.

[ghstack-poisoned]

SplitInfinity commented:
I think this is a step in the right direction but maybe not quite there yet. It is possible to specify bags with constant length using the `offset` version (where the offsets are equal distances apart), and in that case, the reduction is on `dim=0`:
```
e = torch.nn.Embedding(...)
eb = torch.nn.EmbeddingBag(..., mode="sum")
input = ...
offsets = ...
# Me trying to implement EmbeddingBag's offset version with Embedding and sum
out1 = torch.stack([torch.sum(e(input)[offsets[i]:(offsets[i+1] if i != len(offsets)-1 else e(input).shape[0])], dim=0) for i in range(len(offsets))])
out2 = eb(input, offsets)
```

zou3519 commented on torch/nn/modules/sparse.py:
```diff
@@ -188,9 +188,9 @@ class EmbeddingBag(Module):
-        * with ``mode="sum"`` is equivalent to :class:`~torch.nn.Embedding` followed by ``torch.sum(dim=0)``,
-        * with ``mode="mean"`` is equivalent to :class:`~torch.nn.Embedding` followed by ``torch.mean(dim=0)``,
-        * with ``mode="max"`` is equivalent to :class:`~torch.nn.Embedding` followed by ``torch.max(dim=0)``.
+        * with ``mode="sum"`` is equivalent to :class:`~torch.nn.Embedding` followed by ``torch.sum(dim=1)``,
+        * with ``mode="mean"`` is equivalent to :class:`~torch.nn.Embedding` followed by ``torch.mean(dim=1)``,
+        * with ``mode="max"`` is equivalent to :class:`~torch.nn.Embedding` followed by ``torch.max(dim=1)``.
```
I agree with your assessment that this isn't completely correct, because of how EmbeddingBag works with offsets.
1. without offsets: the input must be a 2D tensor and we're taking a sum on the last dimension
2. with offsets: the input must be a 1D tensor with offsets and we're taking a sum on every example in the batch.

We could do two things. The first is to clarify `For bags of constant length and no :attr:`per_sample_weights` and 2D inputs (with no offests)...`. The second is to say something like `mode=sum is equivalent to nn.Embedding followed by a sum of each example in the batch`

SplitInfinity replied:
If we choose to clarify `For bags of constant length and no :attr:per_sample_weights and 2D inputs (with no offsets) ...`, the reduction in that case is still on `dim=1`, right?
```
>>> import torch                                                                                     
>>> weights = torch.nn.Parameter(torch.randn(10, 3))
>>> e = torch.nn.Embedding(10, 3)                  
>>> eb = torch.nn.EmbeddingBag(10, 3, mode="sum")
>>> e.weight = weights                            
>>> eb.weight = weights                                                                              
>>> inputs = torch.LongTensor([[1, 7, 3, 4], [5, 0, 2, 6]])
>>> torch.sum(e(inputs), dim=1)
tensor([[ 0.1180,  0.9364, -0.5337],
        [ 0.7835, -2.5901,  3.0090]], grad_fn=<SumBackward1>)
>>> eb(inputs)
tensor([[ 0.1180,  0.9364, -0.5337],
        [ 0.7835, -2.5901,  3.0090]], grad_fn=<EmbeddingBagBackward>)    
```

SplitInfinity replied:
I added `and 2D inputs` but kept `dim=1`.

Commit:
Update on "[docs] Fix EmbeddingBag docs"


**Summary**
This commit updates the documentation for `EmbeddingBag` to say that for
bags of constant length with no per-sample weights, the class is
equivalent to `Embedding` followed by `torch.sum(dim=1)`. The current
docs say `dim=0` and this is readily falsifiable.

**Test Plan**
1) Tried `Embedding` + `sum` with `dim`=0,1 in interpreter and compared
to `EmbeddingBag`
```
>>> import torch                                                                                     
>>> weights = torch.nn.Parameter(torch.randn(10, 3))              
>>> e = torch.nn.Embedding(10, 3)                                                                    
>>> eb = torch.nn.EmbeddingBag(10, 3, mode="sum")                                                    
>>> e.weight = weights                                                                               
>>> eb.weight = weights                           
# Use 2D inputs because we are trying to test the case in which bags have constant length                                               
>>> inputs = torch.LongTensor([[4,1,2,7],[5,6,0,3]])
>>> eb(inputs)
tensor([[-2.5497, -0.1556, -0.5166],
        [ 2.2528, -0.3627,  2.5822]], grad_fn=<EmbeddingBagBackward>)
>>> torch.sum(e(inputs), dim=0)
tensor([[ 1.6181, -0.8739,  0.8168],
        [ 0.0295,  2.3274,  1.2558],
        [-0.7958, -0.4228,  0.5961],
        [-1.1487, -1.5490, -0.6031]], grad_fn=<SumBackward1>)
>>> torch.sum(e(inputs), dim=1)
tensor([[-2.5497, -0.1556, -0.5166],
        [ 2.2528, -0.3627,  2.5822]], grad_fn=<SumBackward1>)
```
So clearly `torch.sum` with `dim=0` is not correct here.

2) Built docs and viewed in browser.

*Before*
<img width="882" alt="Captura de Pantalla 2020-10-02 a la(s) 12 26 20 p  m" src="https://user-images.githubusercontent.com/4392003/94963035-557be100-04ac-11eb-986c-088965ac3050.png">

*After*
<img width="893" alt="Captura de Pantalla 2020-10-02 a la(s) 12 26 11 p  m" src="https://user-images.githubusercontent.com/4392003/94963001-44cb6b00-04ac-11eb-9b93-dd64f99b0888.png">

**Fixes**
This commit closes #43197.

[ghstack-poisoned]

Commit:
Update on "[docs] Fix EmbeddingBag docs"


**Summary**
This commit updates the documentation for `EmbeddingBag` to say that for
bags of constant length with no per-sample weights, the class is
equivalent to `Embedding` followed by `torch.sum(dim=1)`. The current
docs say `dim=0` and this is readily falsifiable.

**Test Plan**
1) Tried `Embedding` + `sum` with `dim`=0,1 in interpreter and compared
to `EmbeddingBag`
```
>>> import torch                                                                                     
>>> weights = torch.nn.Parameter(torch.randn(10, 3))              
>>> e = torch.nn.Embedding(10, 3)                                                                    
>>> eb = torch.nn.EmbeddingBag(10, 3, mode="sum")                                                    
>>> e.weight = weights                                                                               
>>> eb.weight = weights                           
# Use 2D inputs because we are trying to test the case in which bags have constant length                                               
>>> inputs = torch.LongTensor([[4,1,2,7],[5,6,0,3]])
>>> eb(inputs)
tensor([[-2.5497, -0.1556, -0.5166],
        [ 2.2528, -0.3627,  2.5822]], grad_fn=<EmbeddingBagBackward>)
>>> torch.sum(e(inputs), dim=0)
tensor([[ 1.6181, -0.8739,  0.8168],
        [ 0.0295,  2.3274,  1.2558],
        [-0.7958, -0.4228,  0.5961],
        [-1.1487, -1.5490, -0.6031]], grad_fn=<SumBackward1>)
>>> torch.sum(e(inputs), dim=1)
tensor([[-2.5497, -0.1556, -0.5166],
        [ 2.2528, -0.3627,  2.5822]], grad_fn=<SumBackward1>)
```
So clearly `torch.sum` with `dim=0` is not correct here.

2) Built docs and viewed in browser.

*Before*
<img width="882" alt="Captura de Pantalla 2020-10-02 a la(s) 12 26 20 p  m" src="https://user-images.githubusercontent.com/4392003/94963035-557be100-04ac-11eb-986c-088965ac3050.png">

*After*
<img width="901" alt="Captura de Pantalla 2020-10-05 a la(s) 11 26 51 a  m" src="https://user-images.githubusercontent.com/4392003/95117732-ea294d80-06fd-11eb-9d6b-9b4e6c805cd0.png">


**Fixes**
This commit closes #43197.

[ghstack-poisoned]

zou3519 reviewed (approved):
Seems reasonable
